local-volume/provisioner/pkg/monitor/monitor.go

```diff
@@ -0,0 +1,326 @@
+			// mark PV and send a event
+			err = monitor.markPV(pv, HostPathNotExist, "yes")
+			if err != nil {
+				glog.Errorf("mark PV: %s failed, err: %v", pv.Name, err)
+			}
+			monitor.Recorder.Event(pv, v1.EventTypeWarning, HostPathNotExist, "host path does not exist now")
+			return
+		}
+	}
+
+	// check PV size
+	// TODO: need to support block device
+	capacityByte, err := monitor.VolUtil.GetFsCapacityByte(mountPath)
+	if err != nil {
+		glog.Errorf("Path %q fs stats error: %v", mountPath, err)
+		return
+	}
+	// actually if PV is provisioned by provisioner, the two value must be equal, but the PV may be
+	// created manually, so the PV capacity must not be greater than FS capacity
+	storage := pv.Spec.Capacity[v1.ResourceStorage]
+	if util.RoundDownCapacityPretty(capacityByte) < storage.Value() {
+		// mark PV and send a event
+		err = monitor.markPV(pv, MisMatchedVolSize, "yes")
+		if err != nil {
+			glog.Errorf("mark PV: %s failed, err: %v", pv.Name, err)
+		}
+		monitor.Recorder.Event(pv, v1.EventTypeWarning, MisMatchedVolSize, "PV size does not fit with FS capacity")
+		return
+	}
+
+	// TODO: check volume type ?
+}
+
+// CheckNodeAffinity looks at the PV node affinity, and checks if the node has the same corresponding labels
+// This ensures that we don't mount a volume that doesn't belong to this node
+func CheckNodeAffinity(pv *v1.PersistentVolume, nodeLabels map[string]string) (bool, error) {
+	affinity, err := helper.GetStorageNodeAffinityFromAnnotation(pv.Annotations)
+	if err != nil {
+		return false, fmt.Errorf("error getting storage node affinity: %v", err)
+	}
+	if affinity == nil {
+		return false, nil
+	}
+
+	if affinity.RequiredDuringSchedulingIgnoredDuringExecution != nil {
+		terms := affinity.RequiredDuringSchedulingIgnoredDuringExecution.NodeSelectorTerms
+		glog.V(10).Infof("Match for RequiredDuringSchedulingIgnoredDuringExecution node selector terms %+v", terms)
+		for _, term := range terms {
+			selector, err := helper.NodeSelectorRequirementsAsSelector(term.MatchExpressions)
+			if err != nil {
+				return false, fmt.Errorf("failed to parse MatchExpressions: %v", err)
+			}
+			if !selector.Matches(labels.Set(nodeLabels)) {
+				return false, fmt.Errorf("NodeSelectorTerm %+v does not match node labels", term.MatchExpressions)
+			}
+		}
+	}
+	return true, nil
+}
+
+// markPV marks PV by adding annotation
+func (monitor *Monitor) markPV(pv *v1.PersistentVolume, ann, value string) error {
+
+	// The volume from method args can be pointing to watcher cache. We must not
+	// modify these, therefore create a copy.
+	volumeClone := pv.DeepCopy()
+
+	_, ok := volumeClone.ObjectMeta.Annotations[ann]
+	if ok {
+		glog.V(4).Infof("PV: %s is ready marked using ann: %s", volumeClone.Name, ann)
+		return nil
+	}
+	metav1.SetMetaDataAnnotation(&volumeClone.ObjectMeta, ann, value)
+
+	var err error
+	var newVol *v1.PersistentVolume
+	// Try to update the PV object several times
+	for i := 0; i < UpdatePVRetryCount; i++ {
```

Comment:
maybe we can use a flag to check whether this updating is success?

Reply:
will add event

Reply:
added